src/internal/layout-engine/engine-solution.ts

```diff
@@ -0,0 +1,251 @@
+    for (const moveDirection of PRIORITY_DIRECTIONS) {
+      const move = getOverlapMove(state, overlapId, overlapIssuerId, moveDirection);
+      if (move !== null) {
+        nextMoveSolutions.push([MoveSolutionState.clone(state), move]);
+      }
+    }
+  }
+  return nextMoveSolutions;
+}
+
+// Returns an evaluated move to resolve the given overlap in the given direction or null if such move is not possible.
+function getOverlapMove(
+  state: MoveSolutionState,
+  overlapId: ItemId,
+  overlapIssuerId: ItemId,
+  moveDirection: Direction
+): null | CommittedMove {
+  const userItem = state.grid.getItem(state.moves[0].itemId);
+  const overlapItem = state.grid.getItem(overlapId);
+  const overlapIssuerItem = state.grid.getItem(overlapIssuerId);
+  const overlapMove = getMoveForDirection(overlapItem, overlapIssuerItem, moveDirection);
+
+  // The move position is outside the grid boundaries.
+  if (overlapMove.x < 0 || overlapMove.y < 0 || overlapMove.x + overlapMove.width > state.grid.width) {
+    return null;
+  }
+
+  // Subsequent item overlap moves in the opposite directions do not contribute to solution.
+  const prevOverlapMove = getLastSolutionMove(state, overlapItem.id);
+  if (prevOverlapMove && checkOppositeDirections(prevOverlapMove.direction, moveDirection)) {
+    return null;
+  }
+
+  const pathOverlaps = getPathOverlaps(state, overlapMove, overlapIssuerItem);
+  for (const overlap of pathOverlaps) {
+    // Not allowed to intersect with the user-controlled item.
+    if (overlap.id === userItem.id) {
+      return null;
+    }
+    // Not allowed to intersect with conflicting items.
+    if (state.conflicts?.items.has(overlap.id)) {
+      return null;
+    }
+    // Intersecting with items having unresolved overlaps does not contribute to solution.
+    if (state.overlaps.has(overlap.id)) {
+      return null;
+    }
+  }
+
+  const lastIssuerMove = getLastSolutionMove(state, overlapIssuerItem.id);
+  if (!lastIssuerMove) {
+    throw new Error("Invariant violation: overlap issuer has no associated moves.");
+  }
+  const issuerDirection = lastIssuerMove.direction;
+
+  const isSwap = checkIfSwap(overlapMove, lastIssuerMove);
+  const isDifferentIssuerDirection = moveDirection !== issuerDirection;
+  const isOppositeIssuerDirection = checkOppositeDirections(moveDirection, issuerDirection);
+  const userMoveBoundaries = getUserMoveBoundaries(state);
+  const moveVector = getSolutionMovesVector(state);
+
+  // Swap score penalizes non-swap overlap resolutions in case the direction does not match that of the issuer.
+  const swapPenalty = isSwap ? 0 : 20;
+  const differentDirectionPenalty = !isSwap && isDifferentIssuerDirection ? 10 : 0;
+  const oppositeDirectionPenalty = !isSwap && isOppositeIssuerDirection ? 500 : 0;
+  const swapScore = swapPenalty + differentDirectionPenalty + oppositeDirectionPenalty;
+
+  // Overlaps score penalizes moves that cause additional overlaps.
+  const overlapsScore = pathOverlaps.size * 50;
```

Comment:
This takes into account the number of overlaps only and it is enough to support the existing test cases. However, I see an opportunity for improvement in here. The score can also take the overlap sizes into account - the larger overlaps would have a larger penalty which would make the larger items resolve first and the smaller ones can then take the vacant space. That should have a positive effect on the UX but requires manual testing and some good user stories captured as tests.